rio_tiler/io/cogeo.py

```diff
@@ -119,9 +119,19 @@ def __attrs_post_init__(self):
         if self.post_process is not None:
             self._kwargs["post_process"] = self.post_process
 
-        self.dataset = self.dataset or self._ctx_stack.enter_context(
-            rasterio.open(self.input)
-        )
+        if not self.dataset:
+            dataset = self._ctx_stack.enter_context(rasterio.open(self.input))
```

Comment:
We first open the input file (but register it in the `ctx_stack` so it will be automatically closed when we close the COG instance